grace/evaluation/inference.py

```diff
@@ -1,184 +1,264 @@
+                x=data.x, edge_index=data.edge_index
+            )
+
+            # Process node probs into classes predictions:
+            node_soft = node_x.softmax(dim=1).numpy()
+            node_softmax_preds.extend(node_soft)
+
+            # Process edge probs into classes predictions:
+            edge_soft = edge_x.softmax(dim=1).numpy()
+            edge_softmax_preds.extend(edge_soft)
+
+        # Stack the results & return:
+        return (
+            np.stack(node_softmax_preds, axis=0),
+            np.stack(edge_softmax_preds, axis=0),
+            np.stack(node_labels, axis=0),
+            np.stack(edge_labels, axis=0),
+        )
+
+    def set_node_and_edge_probabilities(self, G: nx.Graph) -> None:
+        """Update predicted node / edge probs by Prediction in-place."""
 
-    def set_node_and_edge_probabilities(self, G: nx.Graph):
-        # Process graph into torch_geometric.data:
         data_batch = dataset_from_graph(
             graph=G, mode="whole", in_train_mode=False
         )
-        results = infer_graph_predictions(self.pretrained_gcn, data_batch)
-        n_probabs, e_probabs, n_pred, e_pred, _, _ = results
+        # Process graph into torch_geometric.data:
+        n_probs, e_probs, _, _ = self.infer_graph_predictions(data_batch)
 
-        # Iterate through each node, storing the variables:
-        for idx, node in G.nodes(data=True):
-            prediction = [int(n_pred[idx].item()), n_probabs[idx].numpy()]
+        # Iterate through each node, storing the 3 class predictions:
+        for n_idx, node in G.nodes(data=True):
+            prediction = Prediction(np.append(n_probs[n_idx], 0.0))
             node[GraphAttrs.NODE_PREDICTION] = prediction
 
         for e_idx, edge in enumerate(G.edges(data=True)):
-            prediction = (int(e_pred[e_idx].item()), e_probabs[e_idx].numpy())
+            prediction = Prediction(np.append(e_probs[e_idx], 0.0))
             edge[-1][GraphAttrs.EDGE_PREDICTION] = prediction
 
-    def visualise_model_performance_on_graph(
-        self, G: nx.Graph, positive_class: int = 1
-    ):
-        # Prep the data & plot them:
-        node_true = np.array(
-            [
-                node[GraphAttrs.NODE_GROUND_TRUTH]
-                for _, node in G.nodes(data=True)
-            ]
+    def visualise_prediction_probs_on_graph(self, G: nx.Graph) -> None:
+        """Visualise predictions as shaded points (nodes) / lines (edges)."""
+        assert all(GraphAttrs.NODE_PREDICTION in G.nodes[n] for n in G.nodes)
+        assert all(GraphAttrs.EDGE_PREDICTION in G.edges[e] for e in G.edges)
+        fig = visualise_node_and_edge_probabilities(G=G)
+        return fig
+
+    def get_predictions_for_entire_batch(
+        self,
+        infer_target_list: list[dict[str]],
+        verbose: bool = False,
+    ) -> tuple[npt.NDArray]:
+        """Processes predictions for list of graph targets."""
+
+        data = [[] for _ in range(6)]
```

Comment:
nomenclature is a bit confusing as to what "data" is

Reply:
Good point - returning & unwrapping the ordered list of lists is confusing & assumes that the order will always be known. I restructured this into a dictionary where the correct list can be accessed by its string key. This is now fixed in #250: 

https://github.com/alan-turing-institute/grace/blob/a7cf4100766c67b99d167fee8ea63207b1fb41e3/grace/evaluation/inference.py#L156-L170

https://github.com/alan-turing-institute/grace/blob/a7cf4100766c67b99d167fee8ea63207b1fb41e3/grace/evaluation/inference.py#L220-L229

Does this seem more intuitive?

Reply:
yes, much better!